### PR TITLE
feat(view): add a pickFeaturesAt method in View

### PR DESCRIPTION
**Pull request by zarov**
This new method comes in addition to `pickObjectsAt`: it looks for all
features under the given screen coordinate, and return it sorted by
layer.

In the same time, `FeatureToolTip` has been greatly improved.

### CODE REVIEW

File: examples/js/plugins/FeatureToolTip.js [hunk 1]
```diff
@@ -1,150 +1,224 @@
...
-                    }
+        var layer;
+        for (var layerId in features) {
+            if (features[layerId].length == 0) {
+                continue;
+            }
+
+            layer = layers[layersId.indexOf(layerId)];
+            if (typeof layer.options.filterGeometries == 'function') {
+                features[layerId] = layer.options.filterGeometries(features[layerId], layer.layer);
```
Comment (gchoqueux) [inline examples/js/plugins/FeatureToolTip.js]: I prefer that the user gives a function `getIdFromProperties(properties)` and that we develop with a filtering function.
May be throw a error if this function isn't exist.

File: examples/js/plugins/FeatureToolTip.js [hunk 2]
```diff
@@ -1,150 +1,224 @@
...
+            content += symb + ' ';
+            content += '</span>';
+            content += (geometry.properties.name || geometry.properties.nom || geometry.properties.description || layer.name || '');
+
+            if (feature.type === itowns.FEATURE_TYPES.POINT) {
+                content += '<br/><span class="coord">long ' + feature.coordinates[0].toFixed(4) + '</span>';
+                content += '<br/><span class="coord">lat ' + feature.coordinates[1].toFixed(4) + '</span>';
+            }
+
+            if (geometry.properties && !options.filterAllProperties) {
```
Comment (gchoqueux) [inline examples/js/plugins/FeatureToolTip.js:103]: `filterAllProperties` could explain this option?

File: src/Core/View.js [hunk 3]
```diff
@@ -636,6 +637,75 @@ class View extends THREE.EventDispatcher {
...
+
+        let precision;
+        const precisions = {
+            PX: radius,
+            M: this.isGlobeView && this.controls ? this.controls.pixelsToMeters(radius) : radius,
+            D: this.isGlobeView && this.controls ? this.controls.pixelsToDegrees(radius) : 0,
+        };
+
+        // Get the mouse coordinates to the correct system
+        const positionVector = new THREE.Vector3();
```
Comment (gchoqueux) [inline src/Core/View.js]: could you instance the `positionVector` and `position` before function?

File: src/Core/View.js [hunk 4]
```diff
@@ -636,6 +637,75 @@ class View extends THREE.EventDispatcher {
...
+        layers = layers.map((l) => {
+            const id = l.isLayer ? l.id : l;
+            result[id] = [];
+            return id;
+        });
+
+        let precision;
+        const precisions = {
+            PX: radius,
+            M: this.isGlobeView && this.controls ? this.controls.pixelsToMeters(radius) : radius,
```
Comment (gchoqueux) [inline src/Core/View.js]: the precision in meter in wrong for planar view.

File: src/Core/View.js [hunk 5]
```diff
@@ -636,6 +637,75 @@ class View extends THREE.EventDispatcher {
...
+        const positionVector = new THREE.Vector3();
+        const mouse = (mouseOrEvt instanceof Event) ? this.eventToViewCoords(mouseOrEvt) : mouseOrEvt;
+        this.getPickingPositionFromDepth(mouse, positionVector);
+        const position = new Coordinates(this.referenceCrs, positionVector);
+
+        // Get the tile corresponding to where the cursor is
+        const tiles = Picking.pickTilesAt(this, mouse, precisions.PX, this.tileLayer);
+
+        for (const t in tiles) {
+            if (!tiles[t].object) {
```
Comment (gchoqueux) [inline src/Core/View.js]: `!tiles[t].object` What is this case?

File: src/Core/View.js [hunk 6]
```diff
@@ -636,6 +637,75 @@ class View extends THREE.EventDispatcher {
...
+                continue;
+            }
+
+            for (const materialLayer of tiles[t].object.material.layers) {
```
Comment (gchoqueux) [inline src/Core/View.js]: could use loop on `layers` and use `material.getLayer(id)` to get `parsedData`

File: src/Core/View.js [hunk 7]
```diff
@@ -636,6 +637,75 @@ class View extends THREE.EventDispatcher {
...
+                }
+
+                for (const texture of materialLayer.textures) {
+                    if (!texture.parsedData) {
+                        continue;
+                    }
+
+                    precision = CRS.isMetricUnit(texture.parsedData.crs) ? precisions.M : precisions.D;
+                    FeaturesUtils.filterFeaturesUnderCoordinate(position, texture.parsedData, precision)
+                        .forEach(f => result[materialLayer.id].push(f));
```
Comment (gchoqueux) [inline src/Core/View.js]: could use `Array.concat`?

**Review by gchoqueux — COMMENTED** [review]
Thanks, the performance is good.
After this PR, could you do a PR to apply this feature on vector tile?

File: examples/js/plugins/FeatureToolTip.js (hunk 1 above)
Comment (zarov) [reply examples/js/plugins/FeatureToolTip.js]: The idea here is that the user could filter the features as he wishes, not necessary by using the id.

File: examples/js/plugins/FeatureToolTip.js (hunk 2 above)
Comment (zarov) [reply examples/js/plugins/FeatureToolTip.js:103]: This is the previous `filterAll` option (see the documentation below)

File: src/Core/View.js (hunk 3 above)
Comment (zarov) [reply src/Core/View.js]: Yes !

File: src/Core/View.js (hunk 4 above)
Comment (zarov) [reply src/Core/View.js]: Yes, I forgot to look at this

File: src/Core/View.js (hunk 5 above)
Comment (zarov) [reply src/Core/View.js]: None, I'm removing it, it something I forgot

File: src/Core/View.js (hunk 7 above)
Comment (zarov) [reply src/Core/View.js]: Ok

File: src/Core/View.js (hunk 6 above)
Comment (zarov) [reply src/Core/View.js]: Done

**Comment by zarov** [comment]
Updated

File: examples/js/plugins/FeatureToolTip.js (hunk 1 above)
Comment (gchoqueux) [reply examples/js/plugins/FeatureToolTip.js]: ok

File: examples/js/plugins/FeatureToolTip.js (hunk 2 above)
Comment (gchoqueux) [reply examples/js/plugins/FeatureToolTip.js:103]: yes, the documentation doesn't give a real explanation.

File: src/Core/View.js (hunk 6 above)
Comment (gchoqueux) [reply src/Core/View.js]: `for (const material of tile.object.material.getLayers(layers)) {`

just rename  `material` to `materialLayer`

File: examples/js/plugins/FeatureToolTip.js (hunk 2 above)
Comment (zarov) [reply examples/js/plugins/FeatureToolTip.js:103]: Yes, updated

File: src/Core/View.js (hunk 6 above)
Comment (zarov) [reply src/Core/View.js]: Done

**Comment by zarov** [comment]
I rebased, and added a commit that refactor the whole meter/degree/pixel system to a more simple and generic one. I also marked some methods in `GlobeControls` as deprecated, as I think we can remove them in the next release (or the one after).

I will add some functional testing to really validate those conversions, but for now it seems to give correct values.

**Comment by zarov** [comment]
Bad news, the `getMetersToPixels` and `getPixelsToMeters` don't return the same value, I'll get back to it.

File: src/Utils/FeaturesUtils.js [hunk 8]
```diff
@@ -144,8 +145,23 @@ export default {
 
         // We can take this shortcut because either Feature and
         // FeatureCollection have an extent property
-        if (features.extent && !features.extent.isPointInside(coordinate, epsilon)) {
-            return result;
+        if (features.extent) {
+            // Special case, because of the way tiles in VectorTileParser are
+            // handled (see Feature2Texture for a similar solution)
+            if (features.extent.crs == 'WMTS:PM') {
+                const ex = new Extent(coordinate.crs, 0, 0, 0, 0);
```
Comment (gchoqueux) [inline src/Utils/FeaturesUtils.js]: could you instance ex before function?

File: src/Utils/FeaturesUtils.js [hunk 9]
```diff
@@ -144,8 +145,23 @@ export default {
 
         // We can take this shortcut because either Feature and
         // FeatureCollection have an extent property
-        if (features.extent && !features.extent.isPointInside(coordinate, epsilon)) {
-            return result;
+        if (features.extent) {
+            // Special case, because of the way tiles in VectorTileParser are
+            // handled (see Feature2Texture for a similar solution)
+            if (features.extent.crs == 'WMTS:PM') {
```
Comment (gchoqueux) [inline src/Utils/FeaturesUtils.js]: it isn't the PM case, it's the case when features have scale or/and translation transformation.

File: src/Core/View.js [hunk 10]
```diff
@@ -636,6 +654,120 @@ class View extends THREE.EventDispatcher {
...
+     * @param {number} [pixels=1] - The size, in pixels, to get in meters.
+     * @param {THREE.Vector3} [screenCoord] - The screen coordinate to get the
+     * projected distance at. By default, this is the middle of the screen.
+     *
+     * @return {number} The projected distance in meters.
+     */
+    getPixelsToMeters(pixels = 1, screenCoord) {
+        const br = this.mainLoop.gfxEngine.renderer.domElement.getBoundingClientRect();
+
+        let _screenCoord = (screenCoord instanceof Event) ? this.eventToViewCoords(screenCoord) : screenCoord;
```
Comment (gchoqueux) [inline src/Core/View.js]: Could you use `const meter = pixel * D / this.camera._preSSE;` instead of?

File: src/Core/Prefab/GlobeView.js [hunk 11]
```diff
@@ -151,6 +152,14 @@ class GlobeView extends View {
 
         return super.addLayer(layer, this.tileLayer);
     }
+
+    getPixelsToDegrees(pixels, screenCoord) {
```
Comment (gchoqueux) [inline src/Core/Prefab/GlobeView.js]: There's no need a specific function for `GlobeView`

**Review by gchoqueux — CHANGES_REQUESTED** [review]
It's work perfectly.

File: src/Core/Prefab/GlobeView.js (hunk 11 above)
Comment (zarov) [reply src/Core/Prefab/GlobeView.js]: Why ?

File: src/Core/View.js (hunk 10 above)
Comment (zarov) [reply src/Core/View.js]: Done

File: src/Utils/FeaturesUtils.js (hunk 9 above)
Comment (zarov) [reply src/Utils/FeaturesUtils.js]: done

File: src/Utils/FeaturesUtils.js (hunk 8 above)
Comment (zarov) [reply src/Utils/FeaturesUtils.js]: done

**Comment by zarov** [comment]
Rebased and updated !

File: src/Controls/GlobeControls.js [hunk 12]
```diff
@@ -1105,6 +1105,9 @@ GlobeControls.prototype.setZoom = function setZoom(zoom, isAnimated) {
  * This function compute the scale of a map
  * @param      {number}  pitch   Screen pitch, in millimeters ; 0.28 by default
  * @return     {number}  The zoom scale.
+ *
+ * @deprecated To remove with pixelsToMeters, pixelsToDegrees and
```
Comment (gchoqueux) [inline src/Controls/GlobeControls.js]: `getScale` will have to be moved to view

File: src/Core/View.js [hunk 13]
```diff
@@ -636,6 +640,123 @@ class View extends THREE.EventDispatcher {
...
+     * @return {number} The distance in meters.
+     */
+    getDistanceFromCamera(screenCoord) {
+        const br = this.mainLoop.gfxEngine.renderer.domElement.getBoundingClientRect();
+
+        let _screenCoord = (screenCoord instanceof Event) ? this.eventToViewCoords(screenCoord) : screenCoord;
+        if (!_screenCoord) {
+            // TODO: This is totally arbitrary, there can be nothing here. Maybe
+            // find a better solution for this.
+            _screenCoord = new THREE.Vector2(br.width / 2, br.height / 2);
```
Comment (gchoqueux) [inline src/Core/View.js]: could you use `screen` instead of new instance.

File: src/Core/View.js [hunk 14]
```diff
@@ -636,6 +640,123 @@ class View extends THREE.EventDispatcher {
...
+        }
+
+        this.getPickingPositionFromDepth(_screenCoord, positionVector);
+        position.setFromVector3(positionVector);
```
Comment (gchoqueux) [inline src/Core/View.js]: why you copy `position` from `positionVector`?

File: src/Utils/FeaturesUtils.js [hunk 15]
```diff
@@ -144,8 +147,33 @@ export default {
...
+            // handled (see Feature2Texture for a similar solution)
+            if (features.translation || features.scale) {
+                ex.crs = coordinate.crs;
+                features.extent.as(coordinate.crs, ex);
+                if (!ex.isPointInside(coordinate, epsilon)) {
+                    return result;
+                }
+
+                if (!features.translation) {
+                    features.translation = new THREE.Vector3();
```
Comment (gchoqueux) [inline src/Utils/FeaturesUtils.js]: could you avoid new instance?

File: src/Core/View.js (hunk 13 above)
Comment (zarov) [reply src/Core/View.js]: done

File: src/Controls/GlobeControls.js (hunk 12 above)
Comment (zarov) [reply src/Controls/GlobeControls.js]: done

File: src/Core/View.js (hunk 14 above)
Comment (zarov) [reply src/Core/View.js]: to be used in `pickFeaturesAt` but I modified that

File: src/Utils/FeaturesUtils.js (hunk 15 above)
Comment (zarov) [reply src/Utils/FeaturesUtils.js]: done

File: src/Controls/GlobeControls.js (hunk 12 above)
Comment (gchoqueux) [reply src/Controls/GlobeControls.js]: use `view.getScale()` inside `GlobeControls.getScale()`

File: src/Utils/FeaturesUtils.js (hunk 15 above)
Comment (gchoqueux) [reply src/Utils/FeaturesUtils.js]: this is useless `features.translation` ans `features.scale` have default values.

**Comment by zarov** [comment]
Rebased and cleaned !

**Review by gchoqueux — APPROVED** [review]

**Pull request merged**